### PR TITLE
GH72 XLIB=Off: fix generated code

### PR DESCRIPTION
**Pull request by davidhubbard**
Here is a potential fix for the XLIB=Off issue (#72).

The comment in vktrace_generate.py, "TODO: need a better way to indicate which extensions should be mapped to which Get*ProcAddr" leads me to believe I could expand this pull request to also do that.

But right now this is just the minimum required to get the XLIB=Off build to compile.

### CODE REVIEW

**Comment by karl-lunarg** [comment]
Thanks for the patches!

We're going to fix the demos/cube issue a little differently.  First demos/cube is downstreamed from https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers, and so we'll fix it there.  Instead of simply removing cube from the build, we can fix the code in cube so that it compiles without Xlib support.

My first glance at the Qt vs Xlib issue in the vktrace_viewer suggests that it may be pretty hard to get it to build with Xlib, and I'm not sure how it ever built.  Not sure yet if we'll try to get the viewer to build with Xlib, or get the entire repo to build without Xlib (the direction of your patches), or both.  We're looking at these things now.

**Comment by davidhubbard** [comment]
Ok, sounds good. This was enough for me to proceed with a vktrace / vkreplay so I'm not blocked on it.

I can rework this on top of a change to demos/cube in LoaderAndValidationLayers.

**Comment by karl-lunarg** [comment]
Did you get vktrace_viewer to compile with your PR in place?

File: vktrace/src/vktrace_extensions/vktracevulkan/vkreplay/vkreplay_vkreplay.cpp [hunk 1]
```diff
@@ -2463,6 +2463,7 @@ VkResult vkReplay::manually_replay_vkCreateXcbSurfaceKHR(packet_vkCreateXcbSurfa
     return replayResult;
 }
 
+#ifdef VK_USE_PLATFORM_XLIB_KHR
```
Comment (davidlunarg) [inline vktrace/src/vktrace_extensions/vktracevulkan/vkreplay/vkreplay_vkreplay.cpp]: Can't put an #ifdef around this function. It needs to be in the build of all platforms in order to support cross-platform playback. For example, vkreplay needs to support playing back on Linux a trace file created on Windows. The file vktrace_platform.h contains the type definitions needed for this cross-platform playback.

There are several more instances of this type of #ifdef in this pull request, I'm omitting comments on those, just marking them.

File: vktrace/src/vktrace_extensions/vktracevulkan/vkreplay/vkreplay_vkreplay.cpp [hunk 2]
```diff
@@ -2463,6 +2463,7 @@ VkResult vkReplay::manually_replay_vkCreateXcbSurfaceKHR(packet_vkCreateXcbSurfa
...
 VkResult vkReplay::manually_replay_vkCreateXlibSurfaceKHR(packet_vkCreateXlibSurfaceKHR* pPacket)
```
Comment (davidlunarg) [inline vktrace/src/vktrace_extensions/vktracevulkan/vkreplay/vkreplay_vkreplay.cpp]: cross-platform

File: vktrace/src/vktrace_extensions/vktracevulkan/vkreplay/vkreplay_vkreplay.cpp [hunk 3]
```diff
@@ -2653,6 +2655,7 @@ VkBool32 vkReplay::manually_replay_vkGetPhysicalDeviceXcbPresentationSupportKHR(
 #endif
 }
 
+#ifdef VK_USE_PLATFORM_XLIB_KHR
```
Comment (davidlunarg) [inline vktrace/src/vktrace_extensions/vktracevulkan/vkreplay/vkreplay_vkreplay.cpp]: cross-platform

File: vktrace/src/vktrace_extensions/vktracevulkan/vkreplay/vkreplay_vkreplay.cpp [hunk 4]
```diff
@@ -2667,12 +2670,14 @@ VkBool32 vkReplay::manually_replay_vkGetPhysicalDeviceXlibPresentationSupportKHR
     m_display->get_window_handle();
     return (m_vkFuncs.real_vkGetPhysicalDeviceXlibPresentationSupportKHR(remappedphysicalDevice, pPacket->queueFamilyIndex, pSurf->dpy, m_display->get_screen_handle()->root_visual));
 #elif defined WIN32
+    # TODO : is this needed with the #ifdef guards around the whole function?
```
Comment (davidlunarg) [inline vktrace/src/vktrace_extensions/vktracevulkan/vkreplay/vkreplay_vkreplay.cpp]: cross-platform

File: vktrace/src/vktrace_extensions/vktracevulkan/vkreplay/vkreplay_vkreplay.h [hunk 5]
```diff
@@ -136,8 +136,10 @@ class vkReplay {
     VkResult manually_replay_vkQueuePresentKHR(packet_vkQueuePresentKHR* pPacket);
     VkResult manually_replay_vkCreateXcbSurfaceKHR(packet_vkCreateXcbSurfaceKHR* pPacket);
     VkBool32 manually_replay_vkGetPhysicalDeviceXcbPresentationSupportKHR(packet_vkGetPhysicalDeviceXcbPresentationSupportKHR* pPacket);
+#ifdef VK_USE_PLATFORM_XLIB_KHR
```
Comment (davidlunarg) [inline vktrace/src/vktrace_extensions/vktracevulkan/vkreplay/vkreplay_vkreplay.h]: cross-platform

**Comment by davidhubbard** [comment]
@karl-lunarg Yes, it did compile and run. Interestingly, running SaschaWillems' triangle demo under vktrace crashed in the nvidia driver...

@davidlunarg That makes sense. Ok, I can revise the patch so it includes these for cross-platform tracing.

**Comment by karl-lunarg** [comment]
@davidhubbard It is easier to handle issues if there are separate Github issues open for each one.  There are three here already.  Would you mind opening a new GitHub VulkanTools issue for the triangle demo crash?   I'd do it myself, but it would be good if you could do it so that you can include information like the version of the nvidia driver, reproduction instructions, etc.

**Comment by karl-lunarg** [comment]
The demos/cube issue was just fixed with https://github.com/LunarG/VulkanTools/commit/fde0fdf578f8322000dd761aacaad3b002856224.
We'll avoid using this commit : https://github.com/LunarG/VulkanTools/pull/75/commits/bde4d190cfd4b7660be840244a8c15c8a071ec21 from this PR.

**Comment by karl-lunarg** [comment]
@davidhubbard  We'll probably proceed by immediately adding a BUILD_VKTRACEVIEWER CMake variable to the CMake files and set its default value to Off.   This will restore normal building (except for the viewer) when the user has Qt5 dev libs installed.  Then we'll work on bringing in the rest of your PR.

**Comment by davidhubbard** [comment]
Yup, sounds good.

**Comment by davidhubbard** [comment]
I am out of town for the rest of the week, and will not be able to repond to messages until Monday Aug 15.

**Comment by davidhubbard** [comment]
Still working on this. To recap:

1. e690cf0 just adds the extension.ifdef field. Pending which way you decide to go @karl-lunarg, I'll leave it in place. If you decide you'd like to be able to build the entire repo without XLIB, I'm working on the suggestions by @davidlunarg -- trace files should work cross-platform. (If you decide to just get VkTraceViewer to compile without XLIB, we can close this pull request.)

1. I reverted bde4d19, I see that 4b6ebbc fixes demos/cube (as part of the fde0fdf merge).

1. With1596c81  VkTraceViewer is off by default on Linux (also part of the fde0fdf merge).

1. I'm going to take a look at SaschaWillems' triangle demo in a separate issue.

**Comment by davidhubbard** [comment]
Done. With f2032a4 and 2c5c049 this pull request can now build `vktraceviewer` with Xlib disabled.

Tested with:

`cmake -H. -Bbuild -DCMAKE_BUILD_TYPE=Debug -DBUILD_WSI_XLIB_SUPPORT=Off -DBUILD_VKTRACEVIEWER=On`

**Comment by karl-lunarg** [comment]
Hi Arda,

Can you and David sort this out and pick up these changes needed to build VT with XLIB Off and VKTRACEVIEWER On?

**Comment by davidhubbard** [comment]
Hi Arda, while I don't want to discard the good context on this pull request, I'll prepare another pull request with just the changes to build VT with XLIB Off and VKTRACEVIEWER On.

**Comment by ardaLunarG** [comment]
Hi David, do you mean close this one and pull req #77 only?

**Comment by davidhubbard** [comment]
Hi Arda, yes! If you like pull req #77 I will close this one and we can move ahead with that one.

**Comment by ardaLunarG** [comment]
Closing this pull req as it is going to be handled on #77

**Pull request closed without merging**